lib/path/tests/prop.sh

```diff
@@ -0,0 +1,186 @@
+#!/usr/bin/env bash
+
+# This file implements property tests for the lib/path/default.nix functions.
+# It generates a lot of random path-like strings and runs the functions on
+# them, checking that the expected laws of the functions hold
+#
+# This file depends on ./generate.awk for random generation
+# and ./normalise.nix for applying the functions to the generated paths
+set -euo pipefail
+shopt -s inherit_errexit
+
+if test -z "${TEST_LIB:-}"; then
+  TEST_LIB="$(cd "$(dirname "${BASH_SOURCE[0]}")/../.."; pwd)"
+fi
+
+tmp="$(mktemp -d)"
+clean_up() {
+  rm -rf "$tmp"
+}
+trap clean_up EXIT
+mkdir -p "$tmp/work"
+cd "$tmp/work"
+
+# Defaulting to a random seed but the first argument can override this
+seed=${1:-$RANDOM}
+echo >&2 "Using seed $seed, use \`lib/path/tests/prop.sh $seed\` to reproduce this result"
+
+# The number of random paths to generate. This specific number was chosen to
+# be fast enough while still generating enough variety to detect bugs.
+count=500
+
+# Set this to 1 or 2 to enable debug output
+debug=0
+
+# Fine tuning of the path generator in ./generate.awk
+# These values were chosen to balance the number of generated invalid paths
+# to the variance in generated paths. Set debug=2 just above to see the paths
+extradotweight=64   # The larger this value, the more dots are generated
+extraslashweight=64 # The larger this value, the more slashes are generated
+extranullweight=16  # The larger this value, the shorter the generated strings
+
+die() {
+    echo >&2 "test case failed: " "$@"
+    exit 1
+}
+
+if [[ "$debug" -ge 1 ]]; then
+    echo >&2 "Generating $count random path-like strings"
+fi
+
+# The awk script generates a stream of null-terminated path-like strings
+# We read entry-by-entry into bash write it to a file and the strings array
+declare -a strings=()
+mkdir -p "$tmp/strings"
+while IFS= read -r -d $'\0' str; do
+    echo -n "$str" > "$tmp/strings/${#strings[@]}"
+    strings+=("$str")
+done < <(awk \
+    -f "$TEST_LIB"/path/tests/generate.awk \
+    -v seed="$seed" \
+    -v count="$count" \
+    -v extradotweight="$extradotweight" \
+    -v extraslashweight="$extraslashweight" \
+    -v extranullweight="$extranullweight")
+
+if [[ "$debug" -ge 1 ]]; then
+    echo >&2 "Trying to normalise the generated path-like strings with Nix"
+fi
+
+# Precalculate all normalisations with a single Nix call. Calling Nix for each
+# string individually would take way too long
+nix-instantiate --eval --strict --json \
+    --arg libpath "$TEST_LIB" \
+    --arg stringsDir "$tmp/strings" \
```

Comment:
```suggestion
    --arg strings "with builtins; map (f: readFile $tmp/strings/${f}) (attrNames (readDir $tmp/strings))" \
```

All that setup is tangential to the actual wrangling of the data and I think it should be done here. Similar from importing `lib` and passing it nicely as an attrset.

Reply:
I don't think this is a good idea because it requires being aware of escaping. I really should've used `--argstr stringsDir` before, because `--argstr` is the only way to pass data safely to Nix in a script. In addition, it allows more easily evaluating the Nix file on its own with `nix-instantiate --eval normalise.nix --argstr stringsDir ...`

Reply:
What's better hinges on what the purpose of that script is. If it's supposed to be evaluated on its own, we should state that and make sure it's clear to readers why there is so much impure boilerplate.

From just reading the code I figured that it could as well be a pure function that does what it's told, and the input is built up in the relevant context, namely the test setup.

I agree the invocation is brittle due to its slight cleverness of mixing shell expansion and Nix string interpolation, but that appears like a reasonable price to pay for clearer separation of concerns - as said, given that's the purpose.

Reply:
While pair programming together, we reached consensus to at least read in all the strings at the top of the `normalise.nix` file